router/proxy/reverseproxy.go

```diff
@@ -84,16 +87,47 @@ func (p *ReverseProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 		return
 	}
 
-	res, err := transport.RoundTrip(outreq)
-	if err != nil {
-		p.logf("router: proxy error: %v", err)
-		rw.WriteHeader(http.StatusServiceUnavailable)
-		rw.Write(serviceUnavailable)
+	ctx, cancel := context.WithCancel(ctx)
+	resultc := make(chan httpResult, 1)
+	clientGone := rw.(http.CloseNotifier).CloseNotify()
+
+	go func() {
+		res, err := transport.RoundTrip(ctx, outreq)
+		defer close(resultc)
+		resultc <- httpResult{res, err}
+	}()
+
+	select {
+	case <-ctx.Done():
+		go throwawayResponse(resultc)
+		return
```

Comment:
this is an undefined situation at this point, should it write a 503 response here? checking `ctx.Done()` here seems unnecessary right now. It's already going to be handled in `transporter`, so it could throw away the result right after the request is canceled?